### ISSUE TITLE
visual studio 2019

### ISSUE THREAD

**Issue by mckl8614** [kcat/openal-soft#396]
Ok so I've built the soft using cmake but now I have no Idea how to link openal soft to visual studio project Can you please outline steps how to link it.

**Comment by kcat** [comment]
I'm not familiar with using MSVC, but I believe it should be like any other project. When you build OpenAL Soft, it'll create a `.lib` file (`OpenAL32.lib`) that your project should link to. You'll also need to add the `include\AL` folder as an include path for your project, so you can include the headers.

**Comment by Raulshc** [comment]
Hello @mckl8614 , Cmake-GUI for windows generates the visual studio project files choosing the generator for that project... Take a look to this guide [https://preshing.com/20170511/how-to-build-a-cmake-based-project/](https://preshing.com/20170511/how-to-build-a-cmake-based-project/) (Section Running cmake-gui)

**Comment by Ted-EE** [comment]
@mckl8614 Hello, I got the similar problem when linking openal soft for visual studio. How you solved the problem? Thanks

**Comment by Ted-EE** [comment]
![build](https://user-images.githubusercontent.com/54616485/82104177-ba0cee00-970d-11ea-8084-065e91151d4d.png)
Hello, @kcat @Raulshc , sorry for disturbing. I just built OpenAL Soft with CMake, but could not find the created OpenAL32.lib file. I would appreciate if you could point out my mistake. Thanks
![cmake](https://user-images.githubusercontent.com/54616485/82104176-b7aa9400-970d-11ea-9ffb-582280245509.png)

**Comment by alexey-lysiuk** [comment]
You need to build created solution with Visual Studio. Press Open Project button in CMake GUI. When it opens, select desired configuration, e.g. Release. Use the corresponding menu to build everything.

**Comment by Ted-EE** [comment]
@alexey-lysiuk Thank you for replying!
I just selected Open Project in CMake GUI and selected Release in MSVC. If I need to build 3dSound.cpp file using HRTF from openal soft, is it alright to create the 3dSound.cpp file in Source Files of OpenAL folder?
![vs](https://user-images.githubusercontent.com/54616485/82118245-c9238880-976d-11ea-8bad-5e124175a570.png)

**Comment by alexey-lysiuk** [comment]
> If I need to build 3dSound.cpp file using HRTF from openal soft, is it alright to create the 3dSound.cpp file in Source Files of OpenAL folder?

No, you need to create a separate project/solution (outside of generated one) that includes OpenAL headers and links with OpenAL library.

**Comment by Ted-EE** [comment]
@alexey-lysiuk 
Thanks for replying!
I just added openal-soft-master\include\AL into C++ -> General -> Additional Include Directories. The next step may require me to link the project with OpenAL32.lib. 
As kcat said,
> When you build OpenAL Soft, it'll create a `.lib` file (`OpenAL32.lib`) that your project should link to.
but I could not find the OpenAL32.lib file from openal-soft-master folder.

**Comment by alexey-lysiuk** [comment]
You need to build OpenAL solution generated by CMake.

**Comment by McSinyx** [comment]
I'm not sure if making CMake generating target would make it any easier, but there's currently [a request on that](https://gitlab.kitware.com/cmake/cmake/issues/20026).  Personally I'm not familiar with CMake enough to get it done though.

**Comment by Ted-EE** [comment]
@alexey-lysiuk Thanks for the replying!
I build the OpenAL solution, and the OpenAL32.lib is generated in the Debug folder of openal-soft-master. However, after I add openal-soft-master\include\AL into C++ -> General -> Additional Include Directories and add openal-soft-master\build\Debug (OpenAL32.dll) into Linker -> General -> Additional Library Directories. The function of HRTF still cannot be called.

![3](https://user-images.githubusercontent.com/54616485/82122843-e2d4c800-978d-11ea-867d-f89970f48366.png)

**Comment by Ted-EE** [comment]
> @alexey-lysiuk Thanks for the replying!
> I build the OpenAL solution, and the OpenAL32.lib is generated in the Debug folder of openal-soft-master. However, after I add openal-soft-master\include\AL into C++ -> General -> Additional Include Directories and add openal-soft-master\build\Debug (OpenAL32.dll) into Linker -> General -> Additional Library Directories. The function of HRTF still cannot be called.
> 
> ![3](https://user-images.githubusercontent.com/54616485/82122843-e2d4c800-978d-11ea-867d-f89970f48366.png)

btw, I have added OpenAL (without soft) from https://www.openal.org/ before for my Visual Studio. Will it conflict with openal-soft-main that is the main reason I cannot call the function from openal-soft-master?

**Comment by kcat** [comment]
> The function of HRTF still cannot be called.

You first need to open a device with `alcOpenDevice`. The `device` pointer you show there is uninitialized, but it needs to be a valid open device handle or that query will fail. You will also need to check that the open device supports the `ALC_SOFT_HRTF` extension (make sure `alcIsExtensionPresent(device, "ALC_SOFT_HRTF")` is true before querying `ALC_NUM_HRTF_SPECIFIERS_SOFT`). Additionally, you need to include `alext.h` to get the extensions' enums and such.

**Comment by Ted-EE** [comment]
Thanks for your replying!
Do I need to delete the OpenAL I downloaded from https://www.openal.org/?
No matter how I changed Target Machine to x86 or x64 in Linker-Advanced, they all got such fail.
![4](https://user-images.githubusercontent.com/54616485/82124421-db65ec80-9796-11ea-9500-6ccace975d08.png)
![5](https://user-images.githubusercontent.com/54616485/82124423-dc971980-9796-11ea-93bd-1c9d191b34df.png)

**Comment by alexey-lysiuk** [comment]
This is not how target architecture is set. Select the one in Visual Studio toolbar that matches built OpenAL library. I guess, what you need is x64.

**Comment by Ted-EE** [comment]
Thanks for your replying!
The default one in my Visual Studio is x86. But when I selected x64, I got 183 errors... I think most of my existing libraries are based on x86. So I think maybe what I should do is to build OpenAL Soft based on x64?
![1](https://user-images.githubusercontent.com/54616485/82126841-3acbf880-97a7-11ea-80fe-6dc65776d4d1.png)

**Comment by alexey-lysiuk** [comment]
Build 32-bit OpenAL then. AFAIK, CMake still doesn’t support multiple architectures in Visual Studio solutions. So, you need to configure and generate again, but select x86 target in CMake GUI this time.

**Comment by Ted-EE** [comment]
Thanks for your replying!

This time I select x86 in CMake GUI, and build OpenAL solution. Finally I got the result 8 succeeded, 1 failed, 1 skipped. The error is:
![2](https://user-images.githubusercontent.com/54616485/82128965-46272000-97b7-11ea-9692-a3c5ccdda539.png)

It is really confused.

**Comment by alexey-lysiuk** [comment]
You have 64-bit Qt installed. If you need OpenAL library only, just disable utilities during CMake configuration step. Look for `ALSOFT_UTILS` option.

**Comment by Ted-EE** [comment]
Thanks!
After I disabled the ALSOFT_UTILS option in CMake GUI. The build succeeded!
![3](https://user-images.githubusercontent.com/54616485/82129640-3100c000-97bc-11ea-8813-539443e7a0b6.png)

**Comment by Ted-EE** [comment]
With all your help, now I have added the CMake generated OpenAL32.lib from openal-soft-master\build\Release into Linker -> General -> Additional Library Directories for my own project (outside the generated one), and added openal-soft-master\include\AL into C++ -> General -> Additional Include Directories. BTW, in my own project, I already included the OpenAL library from https://www.openal.org/ and ALut.lib before. But for my project, it still cannot find 'ALC_NUM_HRTF_SPECIFIERS_SOFT'. Is there any step I miss?
![4](https://user-images.githubusercontent.com/54616485/82129993-390e2f00-97bf-11ea-8208-20ad56952c47.png)
Thanks

**Comment by kcat** [comment]
Did you include `alext.h`? Also, you're not checking what `alcIsExtensionPresent` returns. If it's false, the extension isn't available and the following query will causing an error. You're also opening a device, then calling `alutInit` which also opens a device, causing two separate devices to be opened. If you're want to get the device that Alut opens, you can do:
```
alutInit(NULL, NULL);
ALCdevice *device = alcGetContextsDevice(alcGetCurrentContext());
```

**Comment by Ted-EE** [comment]
Thanks for your help! It really helps!

**Comment by Ted-EE** [comment]
Is it alright to check alcIsExtensionPresent like this?
![5](https://user-images.githubusercontent.com/54616485/82131432-0703c980-97cd-11ea-9acd-a4c08fec12c7.png)

Thanks.

**Comment by kcat** [comment]
It needs a valid device handle for the device you're using, and you need to make sure not to query the HRTF count if it's not supported. But otherwise, like that, yes.

**Comment by Ted-EE** [comment]
I would appreciate it if you show me some examples or tutorials to implement HRTF, I could only find http://api.marmalade.shop/ExampleOpenALHRTF.html which seemed detailed.

Thanks.

**Comment by kcat** [comment]
OpenAL Soft has an HRTF example:
https://github.com/kcat/openal-soft/blob/master/examples/alhrtf.c

To note, you don't necessarily need to do anything for HRTF. If OpenAL Soft can detect a given device is headphones, it will automatically use HRTF. Or if the user has already configured OpenAL Soft to use HRTF, apps will automatically have it without needing to do anything. The extension is only there in case an app wants to implement its own toggle, but it's not strictly necessary to use it to enable HRTF.

**Comment by Ted-EE** [comment]
Thanks for your replying!
So all I need to do is to ensure OpenAL Soft can detect the headphones as device. And as you mentioned before, should I write like this to enable OpenAL Soft to detect my device which is headphones?

> alutInit(NULL, NULL);
> ALCdevice *device = alcGetContextsDevice(alcGetCurrentContext());

Thanks.

**Comment by kcat** [comment]
> And as you mentioned before, first I shoud open a device with alcOpenDevice, then I need to check whether the open device supports the ALC_SOFT_HRTF extension, eventually I should do like this to enable OpenAL Soft to detect my device which is headphones?

No, OpenAL Soft automatically tries to detect if the device is headphones from the system. You don't need to do anything for that.

However, if you want to add an option to enable or disable HRTF regardless of what was detected from the system, you will need a handle for the device in question, to query if the `ALC_SOFT_HRTF` extension is available (and if it is, you can pass the appropriate attributes to `alcCreateContext` or `alcResetDeviceSOFT`).

Since you're using Alut to open the device, you will need to initialize that first, then get the current context and its device, which you can then use to query and reset the device as appropriate. You don't need to open a separate second device.

**Comment by Ted-EE** [comment]
Thanks for your replying!

So the only think I need to do now is to do like
alutInit(NULL, NULL);
ALCdevice *device = alcGetContextsDevice(alcGetCurrentContext());
to initialize, and get the current context and devive. Is it alright?

Thanks.

**Comment by kcat** [comment]
Yes.

**Comment by Ted-EE** [comment]
Thank you for the patient help!
Just got another question. Should I delete the OpenAL lib and dll downloaded from https://www.openal.org/, I am not sure if it will conflict with those of OpenAL Soft. The head files of al, alc come from the OpenAL downloaded from https://www.openal.org/
![1](https://user-images.githubusercontent.com/54616485/82158682-3c69ef00-9881-11ea-8cec-c89e6aaa5e45.png)

The reason I ask such question is because after I did like 
alutInit(NULL, NULL);
ALCdevice *device = alcGetContextsDevice(alcGetCurrentContext());
in my program. It did not show difference among the sound set at (-1.0, 1.0, -1.0), the sound set at (-1.0, 0.0, -1.0) and the sound set at (-1.0, -1.0, -1.0).

**Comment by Ted-EE** [comment]
After I disabled the OpenAL32.lib from OpenAL 1.1 SDK\libs\Win32 and the OpenAL 1.1 SDK\include, the problem still exists. So it may not because of conflict. I guess I might make mistakes in my programming that caused the difference in such elevation difference cannot be recognized. I appreciate if you give any suggestion.
Thanks

**Comment by kcat** [comment]
> The reason I ask such question is because after I did like
> alutInit(NULL, NULL);
> ALCdevice *device = alcGetContextsDevice(alcGetCurrentContext());
> in my program. It did not show difference

That alone won't do anything. That just gets you the `ALCdevice*` that Alut opened, you still need to use it to check for or enable HRTF as desired.

**Comment by Ted-EE** [comment]
Thanks for your help!
I just reviewed some information from http://openal.org/pipermail/openal/2014-December/000275.html. and did like 
void query_hrtf() {
    alutInit(NULL, NULL);
    ALCint hrtfenabled = ALC_FALSE;
    ALCdevice* device = alcGetContextsDevice(alcGetCurrentContext());
    alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
    if (!hrtfenabled) {
        printf("HRTF is in use.");  
    }
}
in my code. But the result only showed "HRTF is in use." without showing any sound.

Thanks

**Comment by kcat** [comment]
The check is inverted, so it prints that if HRTF is *not* in use. You're also not playing any sound, so you won't hear anything (you're also not deinitializing Alut at the end of the function, and it seems weird to open a new device to query HRTF and then let it dangle).

**Comment by Ted-EE** [comment]
Thanks for the patient help!
In my code, 
void play_do(float x, float y, float z, float pitch, float gain) {
    //Source position
    ALfloat SourcePos[] = { x, y, z };

    //Source velocity
    ALfloat SourceVel[] = { 0.0, 0.0, 0.0 };

    ALfloat ListenerPos[] = { 0.0, 0.0, 0.0 };

    ALfloat ListenerVel[] = { 0.0, 0.0, 0.0 };

    ALfloat ListenerOri[] = { 0.0, 0.0, -1.0,  0.0, 1.0, 0.0 };

    alListenerfv(AL_POSITION, ListenerPos);
    alListenerfv(AL_VELOCITY, ListenerVel);
    alListenerfv(AL_ORIENTATION, ListenerOri);

    alutInit(NULL, NULL);
    ALCint hrtfenabled = ALC_FALSE;
    ALCdevice* device = alcGetContextsDevice(alcGetCurrentContext());
    alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
    if (hrtfenabled) {
        printf("HRTF is in use.");
    }
    ALuint source1;
    alGenSources(1, &source1);
    ALuint buffer1 = alutCreateBufferFromFile("do_stretched.wav");
    alSourcei(source1, AL_BUFFER, buffer1);
    alSourcef(source1, AL_PITCH, pitch);
    alSourcef(source1, AL_GAIN, gain);
    alSourcefv(source1, AL_POSITION, SourcePos);
    alSourcefv(source1, AL_VELOCITY, SourceVel);

    alSourcePlay(source1);
    ALint state;
    do {
        alGetSourcei(source1, AL_SOURCE_STATE, &state);
    } while (state == AL_PLAYING);

    alDeleteSources(1, &source1);
    alDeleteBuffers(1, &buffer1);
    alutExit();
}

In fact, it can play sound when I delete
    alutInit(NULL, NULL);
    ALCint hrtfenabled = ALC_FALSE;
    ALCdevice* device = alcGetContextsDevice(alcGetCurrentContext());
    alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
    if (hrtfenabled) {
        printf("HRTF is in use.");
    }
But as I mentioned before, it did not show difference between the sound set at (-1.0, 1.0, -1.0) and the sound set at (-1.0, -1.0, -1.0). So I am thinking of whether I have enabled hrtf or not, and add this paragraph Alut. BTW, I would like to know how to deinitialize Alut at the end.

Thanks.

**Comment by kcat** [comment]
> In fact, it can play sound when I delete
> alutInit(NULL, NULL);
> ALCint hrtfenabled = ALC_FALSE;
> ALCdevice* device = alcGetContextsDevice(alcGetCurrentContext());
> alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
> if (hrtfenabled) {
> printf("HRTF is in use.");
> }

Sounds like Alut was already initialized. You only need to call `alutInit` once (when the app starts).

> But as I mentioned before, it did not show difference between the sound set at (-1.0, 1.0, -1.0) and the sound set at (-1.0, -1.0, -1.0).

Your previous post indicates HRTF was not automatically enabled, meaning the system doesn't recognize the device as headphones, so there wouldn't be a difference.

**Comment by Ted-EE** [comment]
Thanks for the replying!

I was just confused about why I had called alutInit twice before, but after I reviewed http://openal.996291.n3.nabble.com/Why-do-I-need-alut-td2766.html, I seemed to know a bit. 
So when I call alutInit(NULL, NULL), which means I do not need to do like 
ALCdevice *device = alcOpenDevice(NULL);
ALCcontext *context = alcCreateContext(device, NULL);
alcMakeContextCurrent(context);

But it seems that in my latest code, I did not do like ALCdevice *device = alcOpenDevice(NULL); to open another device, which made me confused. I would appreciate it if you point the mistake out.
Additionally, I would like to get some suggestion about how to enable HRTF.

Thanks!

**Comment by Ted-EE** [comment]
Does it mean that I didn't install OpenAL Soft in right way if it showed HRTF was not automatically enabled?

Thanks

**Comment by Ted-EE** [comment]
And just make it clear, I had followed all the steps from https://steamcommunity.com/sharedfiles/filedetails/?id=230642515 before Configure UT3 to use OpenAL Soft. So I have no idea what else should I do to enable HRTF.

**Comment by kcat** [comment]
> So when I call alutInit(NULL, NULL), which means I do not need to do like
> ALCdevice *device = alcOpenDevice(NULL);
> ALCcontext *context = alcCreateContext(device, NULL);
> alcMakeContextCurrent(context);

Correct. `alutInit` opens the default device and sets up a context for you, so once you call it, you're ready to start making AL calls (`alGenSources`, `alSourcei`, etc). You only need to do this once since you only need one device to do things on it.

> Additionally, I would like to get some suggestion about how to enable HRTF.

To enable it in your app, find where you first call `alutInit`, and put this after it:
```
ALCdevice *device = alcGetContextsDevice(alcGetCurrentContext());
if(alcIsExtensionPresent(device, "ALC_SOFT_HRTF"))
{
    LPALCRESETDEVICESOFT alcResetDeviceSOFT = (LPALCRESETDEVICESOFT)alcGetProcAddress(device, "alcResetDeviceSOFT");
    const ALCint attributes[3] = {
        ALC_HRTF_SOFT, ALC_TRUE,
        0
    };
    alcResetDeviceSOFT(device, attributes);

    ALCint hrtfenabled = ALC_FALSE;
    alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
    printf("HRTF is%s enabled\n", hrtfenabled ? "" : " not");
}
else
    fprintf(stderr, "Device \"%s\" does not support ALC_SOFT_HRTF\n", alcGetString(device, ALC_DEVICE_SPECIFIER));
```

> Does it mean that I didn't install OpenAL Soft in right way if it showed HRTF was not automatically enabled?

No, it just means Windows doesn't tell apps that the device is headphones.

**Comment by Ted-EE** [comment]
I am really greatful for the help you provide!

After I put it after the alutInit, it output  
![1](https://user-images.githubusercontent.com/54616485/82260286-e5354e80-9954-11ea-904f-8e8c55c641c0.png)


```
void play_do(float x, float y, float z, float pitch, float gain) {

    alutInit(NULL, NULL);
    ALCdevice* device = alcGetContextsDevice(alcGetCurrentContext());
    if (alcIsExtensionPresent(device, "ALC_SOFT_HRTF"))
    {
        LPALCRESETDEVICESOFT alcResetDeviceSOFT = (LPALCRESETDEVICESOFT)alcGetProcAddress(device, "alcResetDeviceSOFT");
        const ALCint attributes[3] = {
            ALC_HRTF_SOFT, ALC_TRUE,
            0
        };
        alcResetDeviceSOFT(device, attributes);

        ALCint hrtfenabled = ALC_FALSE;
        alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
        printf("HRTF is%s enabled\n", hrtfenabled ? "" : " not");
    }
    else
        fprintf(stderr, "Device \"%s\" does not support ALC_SOFT_HRTF\n", alcGetString(device, ALC_DEVICE_SPECIFIER));

    ALfloat SourcePos[] = { x, y, z };

    //Source velocity
    ALfloat SourceVel[] = { 0.0, 0.0, 0.0 };

    ALfloat ListenerPos[] = { 0.0, 0.0, 0.0 };

    ALfloat ListenerVel[] = { 0.0, 0.0, 0.0 };

    ALfloat ListenerOri[] = { 0.0, 0.0, -1.0,  0.0, 1.0, 0.0 };

    alListenerfv(AL_POSITION, ListenerPos);
    alListenerfv(AL_VELOCITY, ListenerVel);
    alListenerfv(AL_ORIENTATION, ListenerOri);


    ALuint source1;
    alGenSources(1, &source1);
    ALuint buffer1 = alutCreateBufferFromFile("do_stretched.wav");
    alSourcei(source1, AL_BUFFER, buffer1);
    alSourcef(source1, AL_PITCH, pitch);
    alSourcef(source1, AL_GAIN, gain);
    alSourcefv(source1, AL_POSITION, SourcePos);
    alSourcefv(source1, AL_VELOCITY, SourceVel);

    alSourcePlay(source1);
    ALint state;
    do {
        alGetSourcei(source1, AL_SOURCE_STATE, &state);
    } while (state == AL_PLAYING);

    alDeleteSources(1, &source1);
    alDeleteBuffers(1, &buffer1);
    alutExit();
}
```

Now I have no idea why it doesn't support HRTF.

**Comment by Ted-EE** [comment]
I am not sure whether it conflict with openal-soft-1.20.1-bin I downloaded before. For openal-soft-1.20.1-bin, I modified the configuration it as
![2](https://user-images.githubusercontent.com/54616485/82263971-f170da00-995b-11ea-8a9d-04ffeb7d6d43.png)

**Comment by kcat** [comment]
> After I put it after the alutInit, it output

It's using Creative's Generic Software device, not OpenAL Soft. Make sure the OpenAL Soft DLL is with your app's executable, otherwise it will use a system-installed one.

**Comment by Ted-EE** [comment]
I am really grateful for all your selfless help with my code!
The HRTF is enabled finally:
![image](https://user-images.githubusercontent.com/54616485/82266122-bb822480-9960-11ea-9e86-0448e4d9624b.png)
